README.md
# NAME

Module::Implementation - Loads one of several alternate underlying implementations for a module

# VERSION

version 0.08

# SYNOPSIS

    package Foo::Bar;

    use Module::Implementation;

    BEGIN {
        my $loader = Module::Implementation::build_loader_sub(
            implementations => [ 'XS',  'PurePerl' ],
            symbols         => [ 'run', 'check' ],
        );

        $loader->();
    }

    package Consumer;

    # loads the first viable implementation
    use Foo::Bar;

# DESCRIPTION

This module abstracts out the process of choosing one of several underlying
implementations for a module. This can be used to provide XS and pure Perl
implementations of a module, or it could be used to load an implementation for
a given OS or any other case of needing to provide multiple implementations.

This module is only useful when you know all the implementations ahead of
time. If you want to load arbitrary implementations then you probably want
something like a plugin system, not this module.

# API

This module provides two subroutines, neither of which are exported.

## Module::Implementation::build\_loader\_sub(...)

This subroutine takes the following arguments.

- implementations

    This should be an array reference of implementation names. Each name should
    correspond to a module in the caller's namespace.

    In other words, using the example in the ["SYNOPSIS"](#synopsis), this module will look
    for the `Foo::Bar::XS` and `Foo::Bar::PurePerl` modules.

    This argument is required.

- symbols

    A list of symbols to copy from the implementation package to the calling
    package.

    These can be prefixed with a variable type: `$`, `@`, `%`, `&`, or
    `*)`. If no prefix is given, the symbol is assumed to be a subroutine.

    This argument is optional.

This subroutine _returns_ the implementation loader as a sub reference.

It is up to you to call this loader sub in your code.

I recommend that you _do not_ call this loader in an `import()` sub. If a
caller explicitly requests no imports, your `import()` sub will not be run at
all, which can cause weird breakage.

## Module::Implementation::implementation\_for($package)

Given a package name, this subroutine returns the implementation that was
loaded for the package. This is not a full package name, just the suffix that
identifies the implementation. For the ["SYNOPSIS"](#synopsis) example, this subroutine
would be called as `Module::Implementation::implementation_for('Foo::Bar')`,
and it would return "XS" or "PurePerl".

# HOW THE IMPLEMENTATION LOADER WORKS

The implementation loader works like this ...

First, it checks for an `%ENV` var specifying the implementation to load. The
env var is based on the package name which loads the implementations. The
`::` package separator is replaced with `_`, and made entirely
upper-case. Finally, we append "\_IMPLEMENTATION" to this name.

So in our ["SYNOPSIS"](#synopsis) example, the corresponding `%ENV` key would be
`FOO_BAR_IMPLEMENTATION`.

If this is set, then the loader will **only** try to load this one
implementation.

If the env var requests an implementation which doesn't match one of the
implementations specified when the loader was created, an error is thrown.

If this one implementation fails to load then loader throws an error. This is
useful for testing. You can request a specific implementation in a test file
by writing something like this:

    BEGIN { $ENV{FOO_BAR_IMPLEMENTATION} = 'XS' }
    use Foo::Bar;

If the environment variable is _not_ set, then the loader simply tries the
implementations originally passed to `Module::Implementation`. The
implementations are tried in the order in which they were originally passed.

The loader will use the first implementation that loads without an error. It
will copy any requested symbols from this implementation.

If none of the implementations can be loaded, then the loader throws an
exception.

The loader returns the name of the package it loaded.

# AUTHOR

Dave Rolsky <autarch@urth.org>

# COPYRIGHT AND LICENSE

This software is Copyright (c) 2014 by Dave Rolsky.

This is free software, licensed under:

    The Artistic License 2.0 (GPL Compatible)


Changes
{{$NEXT}}

0.09     2014-08-24

- Backed out the Sub::Name change in 0.08. It was pointed out to me by Graham
  Knop that adding an XS dependency for a module that's often used to pick
  between XS and non-XS implementations doesn't work so well.


0.08     2014-08-24

- Subroutines copied from an implementation package into the loading package
  are now renamed using Sub::Name. This causes them to be considered part of
  the loading package, which is important for things like
  namespace::autoclean. Reported by Karen Etheridge. RT #98097.


0.07     2013-07-14

- Require Test::Fatal 0.006+ to avoid test failures. Reported by Salve
  Nilsen. RT #76809.


0.06     2012-02-12

- Require Module::Runtime 0.012, which has a number of useful bug fixes.


0.05     2012-02-09

- Make Test::Taint an optional dependency. This module requires XS, and
  requiring a compiler for Module::Implementation defeats its
  purpose. Reported by Peter Rabbitson. RT #74817.


0.04     2012-02-08

- This module no longer installs an _implementation() subroutine in
  callers. Instead, you can call
  Module::Implementation::implementation_for($package) to get the
  implementation used for a given package.


0.03     2012-02-06

- The generated loader sub now returns the name of the package it loaded.


0.02     2012-02-06

- Removed Test::Spelling from this module's prereqs.


0.01     2012-02-06

- First release upon an unsuspecting world.


t/lib/T/Impl1.pm
package T::Impl1;

use strict;
use warnings;

sub return_42 {
    return 42;
}

sub return_package {
    return __PACKAGE__;
}

sub only_in_impl1 {
    return 69;
}

our $SCALAR = 42;
our @ARRAY  = ( 1, 2, 3 );
our %HASH   = ( key => 'val' );

open MULTI, '>&STDOUT' or die 'Unable to dup STDOUT';
our ( %MULTI, @MULTI );

sub perl_5_6  { 5.6 }
sub perl_5_8  { 5.8 }
sub perl_5_10 { 5.10 }
sub perl_5_14 { 5.14 }
sub perl_5_16 { 5.16 }

our $perl_5_6 = 5.6;
our @perl_5_6 = (5.6);
our %perl_5_6 = ( perl => 5.6 );

1;


t/lib/T/Impl2.pm
package T::Impl2;

use strict;
use warnings;

sub return_42 {
    return 42;
}

sub return_package {
    return __PACKAGE__;
}

our $SCALAR = 42;
our @ARRAY  = ( 1, 2, 3 );
our %HASH   = ( key => 'val' );

open MULTI, '>&STDOUT' or die 'Unable to dup STDOUT';
our ( %MULTI, @MULTI );

sub perl_5_6  { 5.6 }
sub perl_5_8  { 5.8 }
sub perl_5_10 { 5.10 }
sub perl_5_14 { 5.14 }
sub perl_5_16 { 5.16 }

our $perl_5_6 = 5.6;
our @perl_5_6 = (5.6);
our %perl_5_6 = ( perl => 5.6 );

1;


t/lib/T/ImplFails1.pm
package T::ImplFails1;

die 'Error loading something or other';


lib/Module/Implementation.pm
package Module::Implementation;

use strict;
use warnings;

use Module::Runtime 0.012 qw( require_module );
use Try::Tiny;

# This is needed for the benefit of Test::CleanNamespaces, which in turn loads
# Package::Stash, which in turn loads this module and expects a minimum
# version.
unless ( exists $Module::Implementation::{VERSION}
    && ${ $Module::Implementation::{VERSION} } ) {

    $Module::Implementation::{VERSION} = \42;
}

my %Implementation;

sub build_loader_sub {
    my $caller = caller();

    return _build_loader( $caller, @_ );
}

sub _build_loader {
    my $package = shift;
    my %args    = @_;

    my @implementations = @{ $args{implementations} };
    my @symbols = @{ $args{symbols} || [] };

    my $implementation;
    my $env_var = uc $package;
    $env_var =~ s/::/_/g;
    $env_var .= '_IMPLEMENTATION';

    return sub {
        my ( $implementation, $loaded ) = _load_implementation(
            $package,
            $ENV{$env_var},
            \@implementations,
        );

        $Implementation{$package} = $implementation;

        _copy_symbols( $loaded, $package, \@symbols );

        return $loaded;
    };
}

sub implementation_for {
    my $package = shift;

    return $Implementation{$package};
}

sub _load_implementation {
    my $package         = shift;
    my $env_value       = shift;
    my $implementations = shift;

    if ($env_value) {
        die "$env_value is not a valid implementation for $package"
            unless grep { $_ eq $env_value } @{$implementations};

        my $requested = "${package}::$env_value";

        # Values from the %ENV hash are tainted. We know it's safe to untaint
        # this value because the value was one of our known implementations.
        ($requested) = $requested =~ /^(.+)$/;

        try {
            require_module($requested);
        }
        catch {
            require Carp;
            Carp::croak("Could not load $requested: $_");
        };

        return ( $env_value, $requested );
    }
    else {
        my $err;
        for my $possible ( @{$implementations} ) {
            my $try = "${package}::$possible";

            my $ok;
            try {
                require_module($try);
                $ok = 1;
            }
            catch {
                $err .= $_ if defined $_;
            };

            return ( $possible, $try ) if $ok;
        }

        require Carp;
        if ( defined $err && length $err ) {
            Carp::croak(
                "Could not find a suitable $package implementation: $err");
        }
        else {
            Carp::croak(
                'Module::Runtime failed to load a module but did not throw a real error. This should never happen. Something is very broken'
            );
        }
    }
}

sub _copy_symbols {
    my $from_package = shift;
    my $to_package   = shift;
    my $symbols      = shift;

    for my $sym ( @{$symbols} ) {
        my $type = $sym =~ s/^([\$\@\%\&\*])// ? $1 : '&';

        my $from = "${from_package}::$sym";
        my $to   = "${to_package}::$sym";

        {
            no strict 'refs';
            no warnings 'once';

            # Copied from Exporter
            *{$to}
                = $type eq '&' ? \&{$from}
                : $type eq '$' ? \${$from}
                : $type eq '@' ? \@{$from}
                : $type eq '%' ? \%{$from}
                : $type eq '*' ? *{$from}
                : die
                "Can't copy symbol from $from_package to $to_package: $type$sym";
        }
    }
}

1;

# ABSTRACT: Loads one of several alternate underlying implementations for a module

__END__

=head1 SYNOPSIS

  package Foo::Bar;

  use Module::Implementation;

  BEGIN {
      my $loader = Module::Implementation::build_loader_sub(
          implementations => [ 'XS',  'PurePerl' ],
          symbols         => [ 'run', 'check' ],
      );

      $loader->();
  }

  package Consumer;

  # loads the first viable implementation
  use Foo::Bar;

=head1 DESCRIPTION

This module abstracts out the process of choosing one of several underlying
implementations for a module. This can be used to provide XS and pure Perl
implementations of a module, or it could be used to load an implementation for
a given OS or any other case of needing to provide multiple implementations.

This module is only useful when you know all the implementations ahead of
time. If you want to load arbitrary implementations then you probably want
something like a plugin system, not this module.

=head1 API

This module provides two subroutines, neither of which are exported.

=head2 Module::Implementation::build_loader_sub(...)

This subroutine takes the following arguments.

=over 4

=item * implementations

This should be an array reference of implementation names. Each name should
correspond to a module in the caller's namespace.

In other words, using the example in the L</SYNOPSIS>, this module will look
for the C<Foo::Bar::XS> and C<Foo::Bar::PurePerl> modules.

This argument is required.

=item * symbols

A list of symbols to copy from the implementation package to the calling
package.

These can be prefixed with a variable type: C<$>, C<@>, C<%>, C<&>, or
C<*)>. If no prefix is given, the symbol is assumed to be a subroutine.

This argument is optional.

=back

This subroutine I<returns> the implementation loader as a sub reference.

It is up to you to call this loader sub in your code.

I recommend that you I<do not> call this loader in an C<import()> sub. If a
caller explicitly requests no imports, your C<import()> sub will not be run at
all, which can cause weird breakage.

=head2 Module::Implementation::implementation_for($package)

Given a package name, this subroutine returns the implementation that was
loaded for the package. This is not a full package name, just the suffix that
identifies the implementation. For the L</SYNOPSIS> example, this subroutine
would be called as C<Module::Implementation::implementation_for('Foo::Bar')>,
and it would return "XS" or "PurePerl".

=head1 HOW THE IMPLEMENTATION LOADER WORKS

The implementation loader works like this ...

First, it checks for an C<%ENV> var specifying the implementation to load. The
env var is based on the package name which loads the implementations. The
C<::> package separator is replaced with C<_>, and made entirely
upper-case. Finally, we append "_IMPLEMENTATION" to this name.

So in our L</SYNOPSIS> example, the corresponding C<%ENV> key would be
C<FOO_BAR_IMPLEMENTATION>.

If this is set, then the loader will B<only> try to load this one
implementation.

If the env var requests an implementation which doesn't match one of the
implementations specified when the loader was created, an error is thrown.

If this one implementation fails to load then loader throws an error. This is
useful for testing. You can request a specific implementation in a test file
by writing something like this:

  BEGIN { $ENV{FOO_BAR_IMPLEMENTATION} = 'XS' }
  use Foo::Bar;

If the environment variable is I<not> set, then the loader simply tries the
implementations originally passed to C<Module::Implementation>. The
implementations are tried in the order in which they were originally passed.

The loader will use the first implementation that loads without an error. It
will copy any requested symbols from this implementation.

If none of the implementations can be loaded, then the loader throws an
exception.

The loader returns the name of the package it loaded.
